### DISCUSSION TITLE
Dump GBA e-Reader Cards

### DISCUSSION THREAD

**Discussion by DerVerruckteFuchs** [sanni/cartreader#901] [Ideas & Suggestions]
As I am reviewing my collection, I am finding that I'd like to dump my old e-Reader e-Cards. It would be nice to be able to plug in an e-Reader, and dump the e-Card. The format is interesting since some e-Card sets could load games directly onto the e-Reader, and some e-Cards would send data over a GBA link cable to another GBA with a compatible game, or a GBA->GC link cable to a compatible GameCube game. I'm not sure if different reading modes would be necessary, or if they could be dumped directly from the e-Reader's memory.

The three native e-Reader modes I am familiar with (in arbitrary order):
- 1: everything is done directly on the e-Reader for e-Card games
- 2: GBA->GBA connections over serial
- 3: GBA->GC connections over serial
Gen 1 Japanese e-Readers can do mode 1, but not modes 2 and 3 since they lack a GBA link cable port. Western e-Readers and Japanese e-Reader+ (gen 2 models) can do all modes. 

Ideally we'd want to just read an e-Card and dump it over the GBA cart interface. I'm not sure if an unmodded e-Reader would do that, or if we could use DMA on the e-Reader itself to dump the e-Card. With an unmodded GBA and e-Reader, it may be possible to get e-Card dumps over a GBA serial link cable plugged into the cart reader for modes 2 and 3. Mode 1 may be trickier since the e-Reader is not expecting to transmit e-Card data over serial in that mode I think.

The GBA can receive game/demo/software data over the serial port natively. Homebrew can be loaded this way. For example, there is GC/Wii homebrew that uses a GBA->GC link cable to dump GBA ROMS.[[1]](https://github.com/FIX94/gba-link-cable-dumper) The GBA gets sent a program that it will run that can dump the GBA BIOS and any ROMs plugged in over the serial cable. However, ROM dumping this way is pretty slow since the data throughput of the serial cable is pretty low. The GBA can play software sent to it from a GameCube natively. Sonic Adventure 2 used the feature to send the Tiny Chao Garden to GBA (everything would be lost if the GBA was turned off, so any progress needed to be sent back to the GameCube).

What I am thinking, similarly to the GC/Wii based GBA ROM dumper program, is that it may be possible to have an e-Card dumper program sent to the GBA over the serial cable.

Better yet, instead of using the serial cable or GBA at all, an e-Card dumper program printed on some e-Cards that will put the e-Reader itself into a card dumping mode that will communicate with the cart reader. The custom e-Reader code could then communicate with the cart reader to let it know when e-Cards are being dumped or finished dumping, and that e-Card dumping status info could be put on screen. This is assuming that the e-Reader doesn't need all of the GBA hardware to function.

In practicality, as far as the homebrew method is concerned, the serial method may be a bit easier. However, I love the idea of custom e-Cards being used to dump e-Cards.

As I was researching a bit, I found some other git projects that work with GBA e-Readers/e-Cards. I'm not sure how helpful they'll be, but it might help as a starting point.

https://github.com/RollingStar/CardShark
https://github.com/d1x/ereader-card-disassemblies?tab=readme-ov-file
https://github.com/The-Revvy/kirby-1st-recreation

A webpage with more detailed info:
https://www.caitsith2.com/ereader/tech.htm

**Comment by AdmiralCurtiss** [comment]
I randomly found this while searching for an upload of my old flash dumper, and while it's not exactly what you're suggesting, you may still be interested in it: http://www.mediafire.com/file/y0brn9o4o8tuu3v/flash_dumper_v6.7z/file

This is an executable that you can put into the e-Reader save file (it takes up the 'saved game' slot) that can be used to dump cards. They are saved into the e-Reader's save file, and can be read out with any tool that can copy 1M flash save files off a GBA cartridge.
